core/shared/src/main/scala/zio/Semaphore.scala

```diff
@@ -18,153 +18,173 @@ package zio
 
 import zio.stacktracer.TracingImplicits.disableAutoTrace
 import zio.stm.TSemaphore
-
+import java.util.concurrent.atomic.{AtomicLong, AtomicReference}
+import java.util.concurrent.ConcurrentLinkedQueue
 import scala.annotation.tailrec
 import scala.collection.immutable.{Queue => ScalaQueue}
 
-/**
- * An asynchronous semaphore, which is a generalization of a mutex. Semaphores
- * have a certain number of permits, which can be held and released concurrently
- * by different parties. Attempts to acquire more permits than available result
- * in the acquiring fiber being suspended until the specified number of permits
- * become available.
- *
- * If you need functionality that `Semaphore` doesnt' provide, use a
- * [[TSemaphore]] and define it in a [[zio.stm.ZSTM]] transaction.
- */
 sealed trait Semaphore extends Serializable {
-
-  /**
-   * Returns the number of available permits.
-   */
   def available(implicit trace: Trace): UIO[Long]
-
-  /**
-   * Returns the number of tasks currently waiting for permits. The default
-   * implementation returns 0.
-   */
-  def awaiting(implicit trace: Trace): UIO[Long] = ZIO.succeed(0L)
-
-  /**
-   * Executes the specified workflow, acquiring a permit immediately before the
-   * workflow begins execution and releasing it immediately after the workflow
-   * completes execution, whether by success, failure, or interruption.
-   */
+  def awaiting(implicit trace: Trace): UIO[Long]
   def withPermit[R, E, A](zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, A]
-
-  /**
-   * Returns a scoped workflow that describes acquiring a permit as the
-   * `acquire` action and releasing it as the `release` action.
-   */
   def withPermitScoped(implicit trace: Trace): ZIO[Scope, Nothing, Unit]
-
-  /**
-   * Executes the specified workflow, acquiring the specified number of permits
-   * immediately before the workflow begins execution and releasing them
-   * immediately after the workflow completes execution, whether by success,
-   * failure, or interruption.
-   */
   def withPermits[R, E, A](n: Long)(zio: ZIO[R, E, A])(implicit trace: Trace): ZIO[R, E, A]
-
-  /**
-   * Returns a scoped workflow that describes acquiring the specified number of
-   * permits and releasing them when the scope is closed.
-   */
   def withPermitsScoped(n: Long)(implicit trace: Trace): ZIO[Scope, Nothing, Unit]
+  def release(implicit trace: Trace): UIO[Unit] = releaseN(1L)
+  def releaseN(n: Long)(implicit trace: Trace): UIO[Unit]
 }
 
 object Semaphore {
 
-  /**
-   * Creates a new `Semaphore` with the specified number of permits.
-   */
-  def make(permits: => Long)(implicit trace: Trace): UIO[Semaphore] =
-    ZIO.succeed(unsafe.make(permits)(Unsafe.unsafe))
+  def make(permits: => Long, fairness: Boolean = true)(implicit trace: Trace): UIO[Semaphore] =
```

Comment:
Would it make more sense to have a separate constructor for making an unfair semaphore?